crates/bevy_scene/src/dynamic_scene.rs

```diff
@@ -87,17 +97,28 @@ impl DynamicScene {
                         }
                     })?;
 
+                // If this component references entities in the scene, track it
+                // so we can update it to the entity in the world.
+                if registration.data::<ReflectMapEntities>().is_some() {
+                    scene_mappings
+                        .entry(registration.type_id())
+                        .or_insert(Vec::new())
+                        .push(entity);
+                }
+
                 // If the entity already has the given component attached,
                 // just apply the (possibly) new value, otherwise add the
                 // component to the entity.
                 reflect_component.apply_or_insert(entity_mut, &**component);
             }
         }
 
-        for registration in type_registry.iter() {
+        // Updates references to entities in the scene to entities in the world
+        for (type_id, entities) in scene_mappings.into_iter() {
+            let registration = type_registry.get(type_id).unwrap();
```

Comment:
Looks like unwrapping here is fine, since by construction `scene_mappings` is populated by `registration`s that come from the `type_registry`. But it would be a good idea to make the expectation clear (through a comment), so that in the future, say if previous code changes, and the code panic, the poor soul dealing with the code has a chance of understanding why it happens.